pkg/cloud-init/cloud-init.go

```diff
@@ -0,0 +1,376 @@
+			}
+		}
+		return vmCopy, nil
+	default:
+		return vm, errors.New(fmt.Sprintf("Unknown CloudInit type %s", dataSource))
+	}
+}
+
+func ValidateArgs(spec *v1.CloudInitSpec) error {
+	if spec == nil {
+		return nil
+	}
+
+	dataSource := getDataSource(spec)
+	switch dataSource {
+	case dataSourceNoCloud:
+		if spec.NoCloudData == nil {
+			return errors.New(fmt.Sprintf("DataSource %s does not have the required data initialized", dataSource))
+		}
+		if spec.NoCloudData.UserDataBase64 == "" {
+			return errors.New(fmt.Sprintf("userDataBase64 is required for cloudInit type %s", dataSource))
+		}
+		if spec.NoCloudData.MetaDataBase64 == "" {
+			return errors.New(fmt.Sprintf("metaDataBase64 is required for cloudInit type %s", dataSource))
+		}
+	default:
+		return errors.New(fmt.Sprintf("Unknown CloudInit dataSource %s", dataSource))
+	}
+
+	return nil
+}
+
+// Place metadata auto-generation code in here
+func ApplyMetadata(vm *v1.VM) {
+	spec := GetCloudInitSpec(vm)
+	if spec == nil {
+		return
+	}
+
+	namespace := precond.MustNotBeEmpty(vm.GetObjectMeta().GetNamespace())
+	domain := precond.MustNotBeEmpty(vm.GetObjectMeta().GetName())
+
+	switch getDataSource(spec) {
+	case dataSourceNoCloud:
+		// Only autogenerate metadata if user defined metadata does not exist
+		if spec.NoCloudData.MetaDataBase64 != "" {
+			return
+		}
+		// TODO Put local-hostname in MetaData once we get pod DNS working with VMs
+		msg := fmt.Sprintf("instance-id: %s-%s\n", domain, namespace)
+		spec.NoCloudData.MetaDataBase64 = base64.StdEncoding.EncodeToString([]byte(msg))
+	}
+}
+
+func RemoveLocalData(domain string, namespace string) error {
+	domainBasePath := GetDomainBasePath(domain, namespace)
+	return os.RemoveAll(domainBasePath)
+}
+
+func GetCloudInitSpec(vm *v1.VM) *v1.CloudInitSpec {
+	// search various places cloud init spec may live.
+	// at the moment, cloud init only exists on disks.
+	for _, disk := range vm.Spec.Domain.Devices.Disks {
+		if disk.CloudInit != nil {
+			return disk.CloudInit
+		}
+	}
+	return nil
+}
+
+func getDataSource(spec *v1.CloudInitSpec) string {
+	if spec == nil {
+		return ""
+	}
+
+	if spec.NoCloudData != nil {
+		return dataSourceNoCloud
+	}
+	return ""
+}
+
+func GenerateLocalData(domain string, namespace string, spec *v1.CloudInitSpec) error {
+	if spec == nil {
+		return nil
+	}
+
+	err := ValidateArgs(spec)
+	if err != nil {
+		return err
+	}
+
+	domainBasePath := GetDomainBasePath(domain, namespace)
+	err = os.MkdirAll(domainBasePath, 0755)
+	if err != nil {
+		return err
+	}
+
+	switch getDataSource(spec) {
+	case dataSourceNoCloud:
+		metaFile := fmt.Sprintf("%s/%s", domainBasePath, "meta-data")
+		userFile := fmt.Sprintf("%s/%s", domainBasePath, "user-data")
+		iso := fmt.Sprintf("%s/%s", domainBasePath, noCloudFile)
+		isoStaging := fmt.Sprintf("%s/%s.staging", domainBasePath, noCloudFile)
+		userData64 := spec.NoCloudData.UserDataBase64
+		metaData64 := spec.NoCloudData.MetaDataBase64
+
+		os.Remove(userFile)
```

Comment:
It's theoretically possible errors besides "no such file or directory" could occur here. e.g. "permission denied". Worth checking?

Reply:
yeah, I debated about that.

My thought process here is I don't want to to block creating the iso and starting the VM because of being unable to cleanup these temporary files. Even if we can't clean of the temporary files here during the 'Generate' function, they will get cleaned up later on when the entire folder is deleted as a result of the VM shutting down.

so... I could check for errors, but I think I'd just log that they occurred rather than taking any action. I don't feel very strongly one way or the other.

Reply:
+1 on logging. My concern is that if there's a (completely unexpected) issue with cleanup before starting, that doesn't bode well for the rest of the process, so making this issue visible is worth doing.